tests/test_gradient.py

```diff
@@ -1081,3 +1083,70 @@ def test_jacobian_disconnected_inputs():
     func_s = pytensor.function([s2], jacobian_s)
     val = np.array(1.0).astype(pytensor.config.floatX)
     assert np.allclose(func_s(val), np.zeros(1))
+
+
+class TestHessProd:
+    def test_rosen(self):
+        x = vector("x", dtype="float64")
+        rosen = (100 * (x[1:] - x[:-1] ** 2) ** 2 + (1 - x[:-1]) ** 2).sum()
+        p = vector("p")
+
+        rosen_hess_prod_pt = hessian_prod(rosen, wrt=x, p=p)
+
+        x_test = 0.1 * np.arange(9)
+        p_test = 0.5 * np.arange(9)
+        np.testing.assert_allclose(
+            rosen_hess_prod_pt.eval({x: x_test, p: p_test}),
+            rosen_hess_prod(x_test, p_test),
+        )
+
+    def test_multiple_wrt(self):
+        x = vector("x", dtype="float64")
+        y = vector("y", dtype="float64")
+        p_x = vector("p_x", dtype="float64")
+        p_y = vector("p_y", dtype="float64")
+
+        cost = (x**2 - y**2).sum()
+        hessp_x, hessp_y = hessian_prod(cost, wrt=[x, y], p=[p_x, p_y])
+
+        hessp_fn = pytensor.function([x, y, p_x, p_y], [hessp_x, hessp_y])
+        test = {
+            # x, y don't matter
+            "x": np.random.normal(size=(3,)),
+            "y": np.random.normal(size=(3,)),
+            "p_x": [1, 2, 3],
+            "p_y": [3, 2, 1],
+        }
+        hessp_x_eval, hessp_y_eval = hessp_fn(**test)
+        np.testing.assert_allclose(hessp_x_eval, [2, 4, 6])
+        np.testing.assert_allclose(hessp_y_eval, [-6, -4, -2])
+
+    def test_doc_example(self):
+        import numpy as np
+        from scipy.optimize import minimize
+
+        from pytensor import function
+        from pytensor.gradient import hessian_prod, jacobian
+        from pytensor.tensor import vector
+
+        x = vector("x")
+        p = vector("p")
+
+        rosen = (100 * (x[1:] - x[:-1] ** 2) ** 2 + (1 - x[:-1]) ** 2).sum()
+        rosen_hessp = hessian_prod(rosen, x, p)
+        rosen_jac = jacobian(rosen, x)
+
+        rosen_fn = function([x], rosen)
+        rosen_jac_fn = function([x], rosen_jac)
+        rosen_hessp_fn = function([x, p], rosen_hessp)
+        x0 = np.array([1.3, 0.7, 0.8, 1.9, 1.2])
+        res = minimize(
+            rosen_fn,
+            x0,
+            method="Newton-CG",
+            jac=rosen_jac_fn,
+            hessp=rosen_hessp_fn,
+            options={"xatol": 1e-8},
+        )
+        assert res.success
+        np.testing.assert_allclose(res.x, np.ones_like(x0), atol=1e-3)
```

Comment:
@aseyboldt Any idea why Scipy stops at a solution with higher atol error? It seems to work better in their example: https://docs.scipy.org/doc/scipy/tutorial/optimize.html#hessian-product-example

Is our jac / hessp code crap?

Reply:
That function is pretty ugly I think, but that error is suspiciously high... :thinking:

Reply:
Maybe we should check it explicitly against finite diff?
ie something like

```

p = np.random.default_rng(42).normal(size=len(x0))
h = 1e-5  # this should probably depend on the norm of p...

# And there are more stable versions of this...
approximated_hessian_product = (rosen_jac_fn(x0 + h * p) - rosen_jax_fn(x0)) / h
```

Reply:
some more discussion about the finite diff version is here: https://github.com/roualdes/bridgestan/pull/126

Reply:
I'm testing against the scipy hessp function, and that comes back fine (default atol/rtol af assert_allclose`)

Reply:
Seems to be a lie in the scipy docs. If I run it with their functions I don't get any better error

Reply:
Derp, nevermind, I wrote `xatol` instead of `xtol`. Why would scipy check for unused arguments :D

Reply:
Scipy docs are terrible and the arguments are worse